proposals/28-deal-flow-e2e-tests.md

```diff
@@ -0,0 +1,192 @@
+
+#### User workflow example
+<!--
+_How would a developer or user use this new capability?_
+(short paragraph)
+-->
+
+_To be confirmed_. In general it should be runable from CI, and users should be able to initiate ad-hoc runs with particular branches to test changes.
+
+Some ideas:
+
+* Run on every PR to Lotus to identify regressions.
+* Run nightly as a separate project to collect and graph data such as:
+    * Average "time to deal".
+    * Minimum number of commands to store data.
+
+#### Impact
+<!--
+_How would this directly contribute to web3 dev stack product-market fit?_
+
+Explain how this addresses known challenges or opportunities.
+What awesome potential impact/outcomes/results will we see if we nail this project?
+-->
+
+The tests are likely to expose bugs and bottlenecks in the flow. Resolving these issues will improve the user experience.
+
+They will give us a better understanding of the flow and it's shortcomings. This should help to inform future projects looking to streamline the process.
+
+The tests will gather data on the flow that can be used to help future projects determine their success.
+
+#### Leverage
+<!--
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+Explain the opportunity or leverage point for our subsequent velocity/impact (e.g. by speeding up development, enabling more contributors, etc)
+-->
+
+Having adequete E2E tests for deal flow will catch any regressions introduced that are not tested by unit/functional tests. This saves time fixing bugs discovered in production, shields users from experiencing issues, and maintains the view of professionalism of the Lotus team.
+
+#### Confidence
+<!--
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+Explain why this rating
+-->
+
+```
+ICE Score = (Impact + Confidence + Ease) / 3
+ICE Score = (7 + 7 + 8) / 3
+ICE Score = 7.333
+```
+
+**Impact 7** because the project touches on the following areas:
+
+* User Aquuisition - Doesn't Abandon.
+* Activation - Makes a deal first time.
+* Retention - Reducing bad experiences that force the user to try something else.
+* Revenue - Generates revenue.
+
+**Confidence 7** because these will generate test results and benchmarking data.
+
+**Ease 8** an estimate that this will take 3 weeks to get up and running. The time is reltively short because there's no changes to Lotus that need to be made, it can be created as a separate greenfield project.
+
+## Project definition
+#### Brief plan of attack
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+1. Identify test scenarios.
+    * See the "Lotus Release Checklist" for some desired E2E tests to include.
+1. Identify tests that will generate benchmarking data.
+1. Decide on runner (BYO or testground?).
```

Comment:
Does PL have standard "canary" patterns and infrastructure?  For example, in the AWS world I know folks use https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries.html (or roll their own with Lambda).

For TestGround, my total-new-guy understanding is that there is work to be done on test ground itself so it can used within CI.  From Stewards conversations I was listening in on, it sounded like FileCoin has done some of this work, but hasn't fully got it across the line.  Anyways, this sounds like a good area to engage the Stewards on.

Reply:
I don't know enough about testground to comment authoratatively. My understanding is that a cluster would be made available that would be used for this testing. Hopefully we can orchestrate/trigger from CI or do some work to make that possible.

Maybe @nonsense or @raulk can speak to this? 🙏

Reply:
@BigLep @alanshaw here is what we have built on Testground in terms of canaries and end-to-end deals tests:

1. Testground is deployed at Testground as a Service (TaaS) at https://ci.testground.ipfs.team/

2. Automated tests are run on every merge to `master` from https://github.com/filecoin-project/lotus . [Relevant part of .circleci/config.yaml](https://github.com/filecoin-project/lotus/blob/master/.circleci/config.yml#L296:L325) . Unfortunately these tests are rather flaky, but definitely worth review in the realm of "end-to-end tests for deal flow in Lotus" - one of the testplans is doing exactly that. It is based on `TestDealFlow` test in Lotus.

If developers want to give Testground as a Service a try, I can explain the relevant configs so that you trigger test plan on the shared cluster from your local machine - Lotus testplans are rather heavy in terms of memory requirements at the moment, so it might be difficult to run multiple miners/clients on a laptop.

---

Feel free to reach out to me about more questions related to Testground or these test plans!

Reply:
Also worth noting there are some end-to-end deal tests in lotus itself:
https://github.com/filecoin-project/lotus/blob/master/api/test/deals.go